### PR TITLE
Address validation

### PR DESCRIPTION
**Pull request by cvalarida**
**Update:** This PR is ready for review, but it doesn't cover all the validations. I'll make a ticket to cover the rest of the validations and get cracking on it after this gets merged (hopefully!). Once this is merged, adding validations should be as easy as adding a function to an array, so shouldn't take too long.

**Note:** I made a little helper action to get me past the error gate. It's still in this PR, but completely commented out.

### CODE REVIEW

**Comment by cvalarida** [comment]
**Update:** I've now got the validation "engine" up and running, but I still need to write the individual checks. Also, it'll be super easy to add new validations now, so we should be golden ~if~ when requirements change.

Moving forward, I just need to write the validations and resolve the hundred-and-one merge conflicts when @markgreenburg gets his branch merged into master.

File: src/js/letters/components/Address.jsx
Comment (annekainicUSDS) [inline src/js/letters/components/Address.jsx]: Think you can remove this?

File: src/js/letters/containers/AddressSection.jsx
Comment (annekainicUSDS) [inline src/js/letters/containers/AddressSection.jsx]: Minor typo: `validations`

File: src/js/letters/containers/AddressSection.jsx
Comment (annekainicUSDS) [inline src/js/letters/containers/AddressSection.jsx]: So will this just return the first errorMessage that is applicable to this field? But there could be multiple error messages if there are multiple validations?

File: src/js/letters/containers/AddressSection.jsx
Comment (annekainicUSDS) [inline src/js/letters/containers/AddressSection.jsx]: Hm, will this ever be true? Aren't we deleting `militaryStateCode` in the actions? Think we may need to instead check if the `['AE', 'AA', 'AP'].includes(address.state)` or something?

File: src/js/letters/containers/AddressSection.jsx
Comment (annekainicUSDS) [inline src/js/letters/containers/AddressSection.jsx]: Is there a reason we can't run `this.validateField(fieldName, address)` here instead of running the validations on all fields?

File: src/js/letters/utils/validations.js
Comment (annekainicUSDS) [inline src/js/letters/utils/validations.js]: How does this syntax work? How does it know what `input` and `fullAddress` are? This looks like magic!

File: src/js/letters/utils/validations.js
Comment (annekainicUSDS) [inline src/js/letters/utils/validations.js]: I think it should be "ZIP"

File: src/js/letters/utils/validations.js
Comment (annekainicUSDS) [inline src/js/letters/utils/validations.js]: Same here with "ZIP"

**Review by annekainicUSDS — COMMENTED** [review]
Looks great so far! Few questions and comments. Also super helpful comments in the code, that was great!

File: src/js/letters/components/Address.jsx
Comment (markgreenburg) [reply src/js/letters/components/Address.jsx]: ES6's arrow functions don't have their own 'this', so just use the 'this' of the calling function. So basically you don't have to bind the method in the constructor.

Comment (cvalarida) [reply src/js/letters/components/Address.jsx]: Yup, exactly what @markgreenburg said. I just think it's nicer than using the constructor to bind `this` to all the functions. It's really not super important in this case, but when you pass functions as props to children, things get confusing since `this` is perhaps not what you expect. I just like to be consistent with it when I think about it.

File: src/js/letters/components/Address.jsx
Comment (cvalarida) [reply src/js/letters/components/Address.jsx]: There's actually no need to pass any parameters, so...I just nixed them and used the constants we're importing.

File: src/js/letters/components/Address.jsx
Comment (cvalarida) [reply src/js/letters/components/Address.jsx]: So I was looking at the [validation table](https://github.com/department-of-veterans-affairs/vets.gov-team/issues/4110#issuecomment-328640426) and saw that state is not even allowed for international addresses...which I thought was odd, but I was gonna make it and then get somebody who knows better to review the thing when I was done. In the comment for that field, it says "...BGS does not provide a means to store state or province for international addresses..."

I'll make a list of things that I thought were weird and want confirmation on before we call it final.

What I missed, though, was actually hiding the field if USA isn't the country...

File: src/js/letters/components/Address.jsx
Comment (cvalarida) [reply src/js/letters/components/Address.jsx]: I don't actually know if it'll fail the back end, but the table says it's not allowed for international addresses, so I figured we might as well take it out.

File: src/js/letters/components/Address.jsx
Comment (cvalarida) [reply src/js/letters/components/Address.jsx]: Ah! Yes. Because now `this.props.value` actually _is_ `this.props.address`. Good catch!

File: src/js/letters/containers/AddressSection.jsx
Comment (cvalarida) [reply src/js/letters/containers/AddressSection.jsx]: Eagle-eye Anne to the rescue! Thanks. 😄

File: src/js/letters/containers/AddressSection.jsx
Comment (cvalarida) [reply src/js/letters/containers/AddressSection.jsx]: Yes, exactly. So this will return the first validation message, but if they have multiple validation errors, they'll have to be displayed / resolved one at a time. I thought it made more sense than something like "Please enter a zip code and please enter a valid zip code" or something, so we just show one message at a time.

File: src/js/letters/containers/AddressSection.jsx
Comment (cvalarida) [reply src/js/letters/containers/AddressSection.jsx]: This smells like a bad merge...Good catch; I'll look into it.

File: src/js/letters/containers/AddressSection.jsx
Comment (cvalarida) [reply src/js/letters/containers/AddressSection.jsx]: Yeah, I'd prefer to do that, but the reason I have it this way is because if one field changes, it may change the validations on other fields.

For example, we might have a validation that says the city needs to be a valid `militaryPostOfficeTypeCode` (FPO and friends, I think) if it's a military address. So if we set our country to USA and state to AE, but city to San Francisco, there'll be a validation error on the city, but as soon as we change the state to California, we want that validation error to go away, so if we don't validate all the fields, we'd need to touch the city again to reset the validation.

To keep it as performant as possible, I keep track of all the fields that have been modified so we only run validation on them instead of _all_ the fields. I _think_ that'll be okay, but...I haven't really tested it out extensively, if I'm honest. 😳

File: src/js/letters/utils/validations.js
Comment (cvalarida) [reply src/js/letters/utils/validations.js]: In `AddressSection`, `validateField()` passes the validation functions these params:
```js
errorMessage = validations[i](fullAddress[fieldName], fullAddress);
```

All validations functions have to have this signature to work.

Would it be helpful for me to write up some documentation in the comments about this?

File: src/js/letters/containers/AddressSection.jsx
Comment (markgreenburg) [inline src/js/letters/containers/AddressSection.jsx]: I haven't seen them anywhere, I think you'd have to map over the `militaryStateNames` array in `helpers` (but I think you already know this...)

File: src/js/letters/containers/AddressSection.jsx
Comment (markgreenburg) [inline src/js/letters/containers/AddressSection.jsx]: Thought this was already in master? Hm.

Comment (markgreenburg) [reply src/js/letters/containers/AddressSection.jsx]: Oh I guess this came in from the merge, forgot this wasn't rebased in.

Comment (cvalarida) [reply src/js/letters/containers/AddressSection.jsx]: Yeah, it was just part of a merge conflict, so here it is...again. Yay, reruns!

File: src/js/letters/containers/AddressSection.jsx
Comment (cvalarida) [reply src/js/letters/containers/AddressSection.jsx]: Ahh, I didn't realize it was there...not sure if I should use that or not right here, though.

File: src/js/letters/components/Address.jsx
Comment (markgreenburg) [inline src/js/letters/components/Address.jsx]: Are there cases where the country could be undefined and the address type is something other than Military? If so, looks like we're defaulting to setting selectedCountry to the country in the address prop. I take it this maps to an empty country dropdown on the front end?

Do we need the `undefined` check?

Comment (cvalarida) [reply src/js/letters/components/Address.jsx]: Hmmm...I'm actually no sure; I didn't write that bit. Lemme look!

Comment (cvalarida) [reply src/js/letters/components/Address.jsx]: My guess is when we get an address from the api which has a type of `MILITARY`, but no country. We want to select 'USA' on the front end so we can do everything we need to...

@annekainicUSDS to verify.

File: src/js/letters/containers/AddressSection.jsx
Comment (markgreenburg) [inline src/js/letters/containers/AddressSection.jsx]: Could just disable the button if there are error messages instead of touching the `handleSave` logic, right?

Comment (cvalarida) [reply src/js/letters/containers/AddressSection.jsx]: Yeah, that's an option. We're not doing that anywhere else as far as I know, so I'd have to rope design in on that decision. I'll write another ticket for it.

**Review by markgreenburg — APPROVED** [review]
LGTM

File: src/js/letters/components/Address.jsx
Comment (annekainicUSDS) [reply src/js/letters/components/Address.jsx]: @markgreenburg @cvalarida according to the table in this ticket (https://github.com/department-of-veterans-affairs/vets.gov-team/issues/4110), I believe it is possible for `country` to be `undefined`, and for the `addressType` to be `domestic`. In the table it states that `countryName` is not required for domestic addresses. `country` is only required for international addresses. So in order to guarantee the address is military, I think we need to do both of these checks.

However, since we're now pulling the countries from the API, I'm not sure any of this code is needed anymore? I can address that in a separate PR once this is merged.

File: src/js/letters/utils/validations.js
Comment (annekainicUSDS) [reply src/js/letters/utils/validations.js]: Ah okay, yes that helps!

File: src/js/letters/containers/AddressSection.jsx
Comment (annekainicUSDS) [reply src/js/letters/containers/AddressSection.jsx]: Gotcha, that makes sense! We can test more today during QA testing.

File: src/js/letters/components/Address.jsx
Comment (annekainicUSDS) [inline src/js/letters/components/Address.jsx]: I'm still not confident that we're handling this correctly. Even though `state` and `zipCode` aren't required for international addresses, I wonder if sending them will fail validation on the back-end. If not, and if they can be accepted, I think it better to still provide text fields for those questions so that someone can enter that information if they want to. Otherwise, it may not be a valid international address without it. I will add a ticket to check on this separately so it doesn't have to hold up this PR, but would like to return to this question.

File: src/js/letters/components/ErrorableSelect.jsx
Comment (annekainicUSDS) [inline src/js/letters/components/ErrorableSelect.jsx]: Are we doing anything with `dirty` now? Do we know the select will be able to handle this new data structure? Above we're just passing `this.props.value`, if we're changing this to an object we may need to instead pass it `this.props.value.value`.

**Review by annekainicUSDS — COMMENTED** [review]
Testing this locally I still found a couple of issues:
1. If I delete the `city` (which is defaulted to a military city), the `state` and `zipCode` fields are hidden. If I unselect `USA` and reselect it again, the fields appear again, but this obviously isn't a great UX.
2. Updating the `city` to "Cambridge", the `state` to "Massachusetts", and the `zipCode` to "02138" (which is a valid address) fails validation when I try to update.

I'm trying to debug locally on what is causing both of these issues, will update here once I figure it out! But wanted to provide review feedback asap.

**Comment by annekainicUSDS** [comment]
Figured out the cause of the first issue I identified above (where the `state` and `zipCode` fields are being hidden when they shouldn't be):

The reason why they are being hidden when a military address undergoes a change (like deleting the value in the `city` field) is because the `type` is being set to "INTERNATIONAL", which then causes the `state` and `zipCode` fields to be hidden until "USA" is reselected for the `country`. The reason why the `type` is being set to "INTERNATIONAL" is because on this line https://github.com/department-of-veterans-affairs/vets-website/pull/6451/files#diff-23b2fcfb82ec34456090de782cb49c98R145 we're checking if the `country` is not "US" or "USA". When military addresses are sent to us, they are sent without a value for `country`, so they would satisfy this check, therefore setting the type to "INTERNATIONAL".

Proposed fix: In the actions, when an address is sent to us as a "MILITARY" address, we delete the military state and city fields and reset the values for the regular `city` and `state` fields (https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/js/letters/actions/letters.js#L82). At the same time, make sure the `country` is set to "USA".

**Review by annekainicUSDS — APPROVED** [review]
LGTM to merge for now and will fix remaining issues in a separate PRs!

**Pull request merged**
